geps/gep-1651.md

```diff
@@ -0,0 +1,154 @@
+# GEP-1651: GatewayAddress Scope
+
+* Issue: [#1651](https://github.com/kubernetes-sigs/gateway-api/issues/1651)
+* Status: Provisional
+
+(See status definitions [here](overview.md#status).)
+
+## TLDR
+
+Allow users to configure a GatewayAddress so that it is only accessible within
+a specific scope (ie. public/cluster local)
+
+## Goals
+
+- Cluster local gateways can be created in a declarative way `ClusterLocal` and
+  are only accessible within a single cluster
+- Define a `Public` scope for addresses that are accessible to the internet
+- Implementations can define & support their own scopes
+
+## Non-Goals
+
+- Per-request/route scope
+- Not a lightweight service mesh
+
+## Introduction
+
+One of the early feature requests for Knative was the ability to deploy an 
+application using Knative's HTTP routing support, but make it only available
+within the cluster. I want to be able to specify both the "internal" 
+(service.namespace.svc) and "external" (service.namespace.example.com). 
+Gateways using the same GatewayClass on the cluster, but ensure that the 
+"internal" service is only reachable within the cluster. This would greatly 
+simplify deployment for users over the instructions we have today.
+
+The reason why we require an "internal" Gateway is so we can take advantage of 
+layer 7 load balancing for our internal services.
+
+## API
+
+We propose extending the `GatewayAddress` type to allow end-users to specify
```

Comment:
Gateway.Status.Addresses is a bit underspecified.

Is the following valid?

Input:
```
spec: { no address set...}
```

Result:
```
spec: { ... }
status:
  addresses
  - scope: ClusterLocal
    value: 1.2.3.4
  - scope: Public
    value: 5.6.7.8
```

?

Like report the external and internal address.

Reply:
I can make a note about the the default being implementation defined - so in theory an implementation could include 

```yaml
addresses:
- scope: ClusterLocal
  type: IPAddress 
  value: 1.2.3.4
- scope: ClusterLocal
  type: Hostname
  value: some-service.namespace.svc.cluster.local
- scope: Public
  type: IPAddress 
  value: 5.6.7.8
- scope: Public
  type: Hostname
  value: some-public-dns.name.com
```

Reply:
:+1: that is probably what we will do (except probably no Public hostname)

Reply:
Can we use [Node address types](https://pkg.go.dev/k8s.io/kubernetes/pkg/apis/core#NodeAddressType) to help shape `scope`? Users already have a mental model established for these address types. Another option is to deprecate the existing Gateway address types and add support for types that are consistent with Node address types. Since precedence has been established by deprecating `NamedAddressType`, I'm in favor of this option instead of introducing a new concept of `scope`.

Reply:
Another option is to deprecate the existing Gateway address types and add support for types that are consistent with Node address types. Since precedence has been established by deprecating `NamedAddressType`, I'm in favor of this option instead of introducing a new concept of `scope`.

Reply:
This is covered in the GEP document